runtime.Object <-> Unstructured conversion

wojtek-t opened the pull request:
Ref #39017

wojtek-t committed:
Implement runtime.Object <-> map[string]interface{} conversion

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/39418)
<!-- Reviewable:end -->

smarterclayton commented on pkg/runtime/unstructured/converter.go:
```diff
@@ -0,0 +1,234 @@
+}
+
+func (c *Converter) FromUnstructured(u map[string]interface{}, obj runtime.Object) error {
+	return c.fromUnstructured(reflect.ValueOf(u), reflect.ValueOf(obj).Elem())
```
Generally we test for this being a pointer and return an error otherwise.

smarterclayton commented on pkg/runtime/unstructured/converter_test.go:
```diff
@@ -0,0 +1,134 @@
+	"WatchEvent",
+)
+
+func TestRoundTrip(t *testing.T) {
```
Add a benchmark for this test and publish the numbers so we have a rough apples to apples against JSON, conversion, and protobuf (use the same generator patterns there).

wojtek-t replied:
Benchmark done - now it's negligibly faster (like 5%), but I think we have a lot of place for optimization in my code.

smarterclayton commented on pkg/runtime/unstructured/converter_test.go:
```diff
@@ -0,0 +1,134 @@
+limitations under the License.
+*/
+
+package unstructured_test
```
Since runtime is now under apimachinery this test won't work anymore.  You'll probably want to create fake types here for your core testing and do full testing under pkg/api.

smarterclayton commented on pkg/runtime/unstructured/converter.go:
```diff
@@ -0,0 +1,234 @@
+func (c *Converter) structFromUnstructured(sv, dv reflect.Value) error {
+	st, dt := sv.Type(), dv.Type()
+	if st.Kind() != reflect.Map {
+		return fmt.Errorf("Cannot restore struct from: %v", st.Kind())
```
nit: follow the error pattern (messages are lower case), also print the type here so someone can debug.

smarterclayton commented on pkg/runtime/unstructured/converter.go:
```diff
@@ -0,0 +1,234 @@
+}
+
+func fieldNameFromField(field *reflect.StructField) string {
+	jsonTag := field.Tag.Get("json")
```
this is fairly expensive (or was) - I think we should consider caching the field -> tag mapping sooner rather than later once we have a benchmark that tells us how bad this overall.

wojtek-t replied:
Done - in fact this improved performance by ~15%

smarterclayton commented on pkg/runtime/unstructured/converter.go:
```diff
@@ -0,0 +1,234 @@
+func fieldNameFromField(field *reflect.StructField) string {
+	jsonTag := field.Tag.Get("json")
+	if len(jsonTag) == 0 {
+		// FIXME: This should start with small letter.
```
Needs to handle `-` and other special cases.

smarterclayton commented on pkg/runtime/unstructured/converter.go:
```diff
@@ -0,0 +1,234 @@
+			// encode it first.
+			data, err := json.Marshal(sv.Interface())
+			if err != nil {
+				return fmt.Errorf("Error encoding to json")
```
error pattern.

smarterclayton commented on pkg/runtime/unstructured/converter.go:
```diff
@@ -0,0 +1,234 @@
+			if err != nil {
+				return fmt.Errorf("Error encoding to json")
+			}
+			ret := unmarshal.Call([]reflect.Value{reflect.ValueOf(data)})[0].Interface()
```
Call is horribly expensive - use an interface check instead `dv.Interface().(json.Unmarshaller).UnmarshalJSON`.

smarterclayton commented on pkg/runtime/unstructured/converter.go:
```diff
@@ -232,3 +232,176 @@ func (c *Converter) structFromUnstructured(sv, dv reflect.Value) error {
+	if st.Implements(marshalerType) {
+		var data []byte
+		var err error
+		if sv.Kind() == reflect.Ptr && sv.IsNil() {
```
Why isn't this check higher up?

smarterclayton commented on pkg/runtime/unstructured/converter.go:
```diff
@@ -232,3 +232,176 @@ func (c *Converter) structFromUnstructured(sv, dv reflect.Value) error {
+	st, dt := sv.Type(), dv.Type()
+
+	// Check if the object has a custom JSON marshaller/unmarshaller.
+	if st.Implements(marshalerType) {
```
Could also be cached, unsure how expensive this is via reflection.

smarterclayton commented on pkg/runtime/unstructured/converter.go:
```diff
@@ -232,3 +232,176 @@ func (c *Converter) structFromUnstructured(sv, dv reflect.Value) error {
+				return err
+			}
+		}
+		if string(data) == "null" {
```
In the benchmark double check that this doesn't cause an allocation.

smarterclayton replied:
I think it would be better to compare `bytes.Equal(data, []byte("null"))` but the benchmark would tell us.

smarterclayton commented on pkg/runtime/unstructured/converter.go:
```diff
@@ -232,3 +232,176 @@ func (c *Converter) structFromUnstructured(sv, dv reflect.Value) error {
+				result := make(map[string]interface{})
+				err := json.Unmarshal(data, &result)
+				if err != nil {
+					return fmt.Errorf("Error decoding from json: %v", err)
```
Need a lot better errors here for most of these cases.

smarterclayton commented on pkg/runtime/unstructured/converter.go:
```diff
@@ -232,3 +232,176 @@ func (c *Converter) structFromUnstructured(sv, dv reflect.Value) error {
+		if string(data) == "null" {
+			// We're done - we don't need to store anything.
+		} else {
+			if len(data) > 0 && data[0] == '"' {
```
Turn this if else if into a switch.

wojtek-t replied:
done

smarterclayton commented on pkg/runtime/unstructured/converter.go:
```diff
@@ -232,3 +232,176 @@ func (c *Converter) structFromUnstructured(sv, dv reflect.Value) error {
+
+func (c *Converter) mapToUnstructured(sv, dv reflect.Value) error {
+	st, dt := sv.Type(), dv.Type()
+	if dt.Kind() == reflect.Interface && dv.NumMethod() == 0 {
```
Seems questionable - why not compare dt against a package variable that holds `reflect.TypeOf(interface{}{})`?

smarterclayton commented on pkg/runtime/unstructured/converter.go:
```diff
@@ -232,3 +232,176 @@ func (c *Converter) structFromUnstructured(sv, dv reflect.Value) error {
+func (c *Converter) mapToUnstructured(sv, dv reflect.Value) error {
+	st, dt := sv.Type(), dv.Type()
+	if dt.Kind() == reflect.Interface && dv.NumMethod() == 0 {
+		dv.Set(reflect.MakeMap(mapStringInterfaceType))
```
Probably should do this lazily.

smarterclayton commented on pkg/runtime/unstructured/converter.go:
```diff
@@ -232,3 +232,176 @@ func (c *Converter) structFromUnstructured(sv, dv reflect.Value) error {
+		return fmt.Errorf("Cannot convert struct to: %v", dt.Kind())
+	}
+
+	if !st.Key().AssignableTo(dt.Key()) && !st.Key().ConvertibleTo(dt.Key()) {
```
Probably should cache and check this elsewhere - do a feasibility check for each st -> dt and cache the results.  This gets expensive.

smarterclayton replied:
Note that cache wouldn't be safe if input is unbounded, so we'd have to be careful to only cache `Go types` -> `map`, and never let the left side be a map[string]interface (which is possible if someone implements runtime.Object over map[string]interface{}).

smarterclayton commented on pkg/runtime/unstructured/converter.go:
```diff
@@ -232,3 +232,176 @@ func (c *Converter) structFromUnstructured(sv, dv reflect.Value) error {
+	}
+
+	for _, key := range sv.MapKeys() {
+		value := reflect.New(dt.Elem()).Elem()
```
This would really be better if it were lazy.

smarterclayton commented on pkg/runtime/unstructured/converter.go:
```diff
@@ -232,3 +232,176 @@ func (c *Converter) structFromUnstructured(sv, dv reflect.Value) error {
+}
+
+func (c *Converter) interfaceToUnstructured(sv, dv reflect.Value) error {
+	return fmt.Errorf("Interface conversion unsupported")
```
include the source and destination type.

smarterclayton commented on pkg/runtime/unstructured/converter.go:
```diff
@@ -232,3 +232,176 @@ func (c *Converter) structFromUnstructured(sv, dv reflect.Value) error {
+					return err
+				}
+				dv.SetMapIndex(reflect.ValueOf(fieldName), subv)
+			} else {
```
I'd invert this condition to:

```
if !fv.IsValid() {
  // no source field, skip
  continue
}
...
```

which reads better and avoids indentation.

smarterclayton commented on pkg/runtime/unstructured/converter.go:
```diff
@@ -232,3 +232,176 @@ func (c *Converter) structFromUnstructured(sv, dv reflect.Value) error {
+			// This field is inlined.
+			if err := c.toUnstructured(fv, dv); err != nil {
+				return err
+			}
```
Would recommend dropping the else and putting a continue here.

smarterclayton commented on pkg/runtime/unstructured/converter.go:
```diff
@@ -232,3 +232,176 @@ func (c *Converter) structFromUnstructured(sv, dv reflect.Value) error {
+	mapStringInterfaceType = reflect.TypeOf(map[string]interface{}{})
+)
+
+func (c *Converter) ToUnstructured(obj runtime.Object, u *map[string]interface{}) error {
```
I'm not sure what value having the `Converter` type brings.  I'd recommend all the private methods just become private functions, which dramatically simplifies this unless there's something usefel you'll be adding to Converter.  All of the leaf methods that don't have a reason should just be functions.

liggitt commented on pkg/runtime/unstructured/converter.go:
```diff
@@ -0,0 +1,419 @@
+	marshalerType          = reflect.TypeOf(new(encodingjson.Marshaler)).Elem()
+	unmarshalerType        = reflect.TypeOf(new(encodingjson.Unmarshaler)).Elem()
+	mapStringInterfaceType = reflect.TypeOf(map[string]interface{}{})
+	// TODO: Can we somehow avoid locking?
```
can use the ReadMostly approach with atomic Load() - https://golang.org/pkg/sync/atomic/#example_Value_readMostly

liggitt commented on pkg/runtime/unstructured/converter.go:
```diff
@@ -0,0 +1,419 @@
+	jsonTag := typeField.Tag.Get("json")
+	if len(jsonTag) == 0 {
+		// FIXME: This should start with small letter.
+		name = typeField.Name
```
shouldn't this be an error?

liggitt commented on pkg/runtime/unstructured/converter.go:
```diff
@@ -0,0 +1,419 @@
+					return fmt.Errorf("error decoding from json: %v", err)
+				}
+				dv.Set(reflect.ValueOf(result))
+			default:
```
what about arrays?

liggitt replied:
and bools?

smarterclayton commented on pkg/runtime/unstructured/BUILD:
```diff
@@ -0,0 +1,36 @@
+package(default_visibility = ["//visibility:public"])
```
Don't forget package needs to move to apimachinery/runtime/unstructured and test case will need to be in pkg/api (the one that references the API types).

wojtek-t replied:
I factored out #40144

smarterclayton commented on pkg/runtime/unstructured/converter.go:
```diff
@@ -0,0 +1,444 @@
+			// No fource field, skip.
+			continue
+		}
+		subv := reflect.New(dt.Elem()).Elem()
```
I think you can avoid a substantial number of allocations if you check for the zero value here before calling reflect.New.  Same for map and pointers.

wojtek-t replied:
Isn't the above if doing exactly this? I mean fv.IsValid is exactly checking for zero value IIUC.

Though I'm not entirely sure what we are allocating here. If dv is map[string]interface{}, then what we are creating here? Maybe this can be optimized somehow, but I need to dig what can we do here.

smarterclayton replied:
IsValid() doesn't check for the zero value, it checks for the zero Value (subtle difference).

```
fmt.Printf("value: %t\n", reflect.ValueOf(int(0)).IsValid())
```

prints "value: true"

smarterclayton commented:
What's interesting to me is ToUnstructured looks like it's twice as expensive as FromUnstructured, which is very surprising (decode is far more expensive than encode for json).  It's because it allocates twice as many objects.

smarterclayton commented on pkg/runtime/unstructured/converter.go:
```diff
@@ -0,0 +1,444 @@
+func sliceToUnstructured(sv, dv reflect.Value) error {
+	dt := dv.Type()
+	if dt.Kind() == reflect.Interface && dv.NumMethod() == 0 {
+		dv.Set(reflect.MakeSlice(reflect.SliceOf(dt), sv.Len(), sv.Cap()))
```
Avoid allocating here if the slice is nil.  If the slice is empty... I think we can keep a cache of empty slices and use them (although we'd need to double check a go slice allocation from an empty slice creates a new backing store).

wojtek-t commented on pkg/runtime/unstructured/converter.go:
```diff
@@ -0,0 +1,472 @@
+/*
```
@smarterclayton - I tried a bunch of other optimizations but most of them were useless, so I removed them as they were complicating code. In the current shape we have:

wojtekt@wojtekt-work:~/go/src/k8s.io/kubernetes$ go test ./pkg/runtime/unstructured -test.run=XXXXX -bench=ToFrom -benchmem -memprofile mem.out -memprofilerate=1
BenchmarkToFromUnstructured-12           	    2000	   1318675 ns/op	   15716 B/op	     295 allocs/op
BenchmarkToFromUnstructuredViaJSON-12    	    1000	   2055132 ns/op	   29659 B/op	     457 allocs/op
PASS
ok  	k8s.io/kubernetes/pkg/runtime/unstructured	5.502s
wojtekt@wojtekt-work:~/go/src/k8s.io/kubernetes$ go test ./pkg/runtime/unstructured -test.run=XXXXX -bench=ToFrom
BenchmarkToFromUnstructured-12           	   10000	    122756 ns/op
BenchmarkToFromUnstructuredViaJSON-12    	    5000	    206080 ns/op
PASS
ok  	k8s.io/kubernetes/pkg/runtime/unstructured	2.484s

Which is not ideal, but I think worth merging. We can probably try optimizing further later.

Can you please take a look?

Also, I need some guidance where I should put the code now with all these moves to separate repositories. Should it go to staging/ directory? If so, how/when it will be available for use e.g. by test in pkg/api ?

smarterclayton replied:
Yeah, should go to staging/src/k8s.io/apimachinery/pkg/conversion/unstructured (since this is about conversion).  That is symlinked into vendor, so your type will be available at `k8s.io/apimachinery/pkg/conversion/unstructured`.

smarterclayton replied:
I had hoped to get more than a 2x speedup but I'm ok with starting as is.  I would like a follow up opened to continue looking at this (I expected to see a 5x speedup).

wojtek-t commented:
@smarterclayton - I cleaned up the PR, PTAL (I added two minor optimizations, but that didn't change much)

@smarterclayton @mikedanese - I have some issues with bazel, which I don't know how to fix. Basically bazel is failing, even though running hack/update-bazel.sh doesn't change anything. Any thoughts how to fix it?

smarterclayton commented on staging/src/k8s.io/apimachinery/pkg/conversion/unstructured/converter_test.go:
```diff
@@ -0,0 +1,152 @@
+		{
+			obj: &C{
+				A: []A{
+					{
```
Test floating point numbers, empty array, array with nil, empty map, nil map, array of interface{} with nils, array of interface with empty arrays, array of booleans, array of numbers, map of numbers.  Would like close to 100% testing on possible types.

wojtek-t replied:
Added few more - didin't add the two:
- array of interface{} with nils
- array of interface with empty arrays,

Those currently do not work (and it was pretty explicit), because the the conversion to from "interface{}" fields is not supported. However, we are not using it anywhere in API and it doesn't seem to be a problem...

smarterclayton replied:
we allow nulls in certain arrays.  So I'm fairly certain it's possible to get an object that generates them.  So I would expect unstructured To -> object to have to support them.  I don't care if it's in the round tripper, but there needs to be a test for those two because it is possible for patch or location to generate that.

smarterclayton replied:
So for instance, if a user sends a patch that contains `Containers: [nil, nil, nil]`, unstructured From will get called on `[]interface{nil, nil, nil}`

wojtek-t replied:
@smarterclayton - I don't understand. Isn't the "I: " field below (in this and next case) testing what you wanted?

wojtek-t reviewed (commented):
@smarterclayton PTAL

smarterclayton commented:
> array of interface{} with nils
> array of interface with empty arrays,
> Those currently do not work (and it was pretty explicit), because the the conversion to from "interface{}" fields is not supported. However, we are not using it anywhere in API and it doesn't seem to be a problem...

JSON imports `["a", true, 1]` as `[]interface{}{"a", true, 1}`, and `[null, null, null]` as `[]interface{}{nil, nil, nil}`

spxtr commented:
@k8s-bot bazel test this
#40526

wojtek-t commented:
>> array of interface{} with nils
>> array of interface with empty arrays,
>> Those currently do not work (and it was pretty explicit), because the the conversion to from "interface{}" fields is not supported. However, we are not using it anywhere in API and it doesn't seem to be a problem...

> JSON imports `["a", true, 1]` as `[]interface{}{"a", true, 1}`, and `[null, null, null]` as `[]interface{}{nil, nil, nil}`

Hmm, but that would mean that I need to add to my type a field like:
```
field []interface{}
```
to my struct. Which is explicitly not supported by the current code.
I think we should focus on JSON that much, as this conversion is not working on JSON - it's working on runtime.Objects, where I don't think it makes sense to have []interface{} field (or sth like that).

Or am I completely missing something? @smarterclayton

smarterclayton commented:
You will have to make patch work with third party resource, which is
unstructured, which could have {nil, nil, nil} I believe.

On Thu, Jan 26, 2017 at 3:47 PM, Wojciech Tyczynski <
notifications@github.com> wrote:

> array of interface{} with nils
> array of interface with empty arrays,
> Those currently do not work (and it was pretty explicit), because the the
> conversion to from "interface{}" fields is not supported. However, we are
> not using it anywhere in API and it doesn't seem to be a problem...
>
> JSON imports ["a", true, 1] as []interface{}{"a", true, 1}, and [null,
> null, null] as []interface{}{nil, nil, nil}
>
> Hmm, but that would mean that I need to add to my type a field like:
>
> field []interface{}
>
> to my struct. Which is explicitly not supported by the current code.
> I think we should focus on JSON that much, as this conversion is not
> working on JSON - it's working on runtime.Objects, where I don't think it
> makes sense to have []interface{} field (or sth like that).
>
> Or am I completely missing something? @smarterclayton
> <https://github.com/smarterclayton>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/39418#issuecomment-275508637>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p-5CoBvN-eBJkF-8HFU3AOoMqx3rks5rWQZHgaJpZM4Laj1V>
> .
>

wojtek-t commented:
> You will have to make patch work with third party resource, which is
unstructured, which could have {nil, nil, nil} I believe.

TPR is a good point. Though I'm not 100% we will hit this path though (do we support strategic merge patch for TPRs? I mean, does it work now?)

Anyway - I think I roughly made it work, or at least the tests that you suggested are now passing :)

@smarterclayton PTAL

wojtek-t commented:
@k8s-bot gci gce test this, issue: #IGNORE (there are problems in head #40592 is trying to fix it)

wojtek-t commented:
@k8s-bot gci gke test this, issue: #IGNORE (up failed)

smarterclayton commented on staging/src/k8s.io/apimachinery/pkg/conversion/unstructured/converter_test.go:
```diff
@@ -0,0 +1,181 @@
+				F: []bool{true, false, false},
+				G: []int{1, 2, 5},
+				H: 3.3,
+				I: []interface{}{nil, nil, nil},
```
You also need a one way test for unstructuredTo that takes these values (round trip only addresses half of the problem).  You need "unrecognized interfaces -> structs" as well for this test to be complete.  I.e. create TestUnstructuredTo and verify the other types I described (anything that could be in valid JSON to map[string]interface{} to a struct).  And that test should do map[string]interface{} -> unstructuredTo -> struct and also map[string]interface{} -> JSON.marshal -> unmarshal struct and compare that they are equal.  You'll want to also verify that you handle mixed mode errors the same (i.e. [3.0, "3.0", nil] should behave the same in unstructuredTo that it does in JSON).

wojtek-t replied:
OK - I added some more tests (hopefully I understood you correctly in what you wanted).

PTAL

smarterclayton commented:
Future improvements:

* It looks like you have a hotspot on maps - this looks disproportionate to the types you have and so i expect it's labels / resources / selectors / annotations which are fairly easy to optimize or there's something else wrong:

```
         0     0% 77.89%    2875433 31.08%  reflect.Value.MapIndex
         0     0% 77.89%    2585090 27.94%  reflect.Value.SetMapIndex
         0     0% 77.89%    2281067 24.65%  k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/conversion/unstructured.sliceToUnstructured
         0     0% 77.89%    2162716 23.37%  reflect.Value.assignTo
```

It might be map[string]string converted from map[string]interface{}, and that instead of doing a generic iteration with reflect, you should be special casing map[string]string (when it's the destination type) with a special for loop that sets a map[string]string).  That might be a big win.

* There are way too many new object creations - I count 260 allocations for ToUnstructured and 130 for FromUnstructured for the first object in the benchmarkItems list.  I tweaked to get that for only this type:

```
{
  "kind": "Pod",
  "apiVersion": "v1",
  "metadata": {
    "name": "etcd-server-e2e-test-wojtekt-master",
    "namespace": "default",
    "selfLink": "/api/v1/namespaces/default/pods/etcd-server-e2e-test-wojtekt-master",
    "uid": "a671734a-e8e5-11e4-8fde-42010af09327",
    "resourceVersion": "22",
    "creationTimestamp": "2015-04-22T11:49:36Z",
    "annotations": {
      "kubernetes.io/config.mirror": "mirror",
      "kubernetes.io/config.source": "file"
    }
  },
  "spec": {
    "volumes": [
      {
        "name": "varetcd",
        "hostPath": {
          "path": "/mnt/master-pd/var/etcd"
        }
      }
    ],
    "containers": [
      {
        "name": "etcd-container",
        "image": "gcr.io/google_containers/etcd:2.0.9",
        "command": [
          "/usr/local/bin/etcd",
          "--addr",
          "127.0.0.1:2379",
          "--bind-addr",
          "127.0.0.1:2379",
          "--data-dir",
          "/var/etcd/data"
        ],
        "ports": [
          {
            "name": "serverport",
            "hostPort": 2380,
            "containerPort": 2380,
            "protocol": "TCP"
          },
          {
            "name": "clientport",
            "hostPort": 2379,
            "containerPort": 2379,
            "protocol": "TCP"
          }
        ],
        "resources": {},
        "volumeMounts": [
          {
            "name": "varetcd",
            "mountPath": "/var/etcd"
          }
        ],
        "terminationMessagePath": "/dev/termination-log",
        "imagePullPolicy": "IfNotPresent"
      }
    ],
    "restartPolicy": "Always",
    "dnsPolicy": "ClusterFirst",
    "nodeName": "e2e-test-wojtekt-master"
  },
  "status": {
    "phase": "Running",
    "conditions": [
      {
        "type": "Ready",
        "status": "True",
        "lastProbeTime": null,
        "lastTransitionTime": null
      }
    ],
    "containerStatuses": [
      {
        "name": "etcd-container",
        "state": {
          "running": {
            "startedAt": "2015-04-22T11:49:32Z"
          }
        },
        "lastState": {},
        "ready": true,
        "restartCount": 0,
        "image": "gcr.io/google_containers/etcd:2.0.9",
        "imageID": "docker://b6b9a86dc06aa1361357ca1b105feba961f6a4145adca6c54e142c0be0fe87b0",
        "containerID": "docker://3cbbf818f1addfc252957b4504f56ef2907a313fe6afc47fc75373674255d46d"
      }
    ]
  }
}
```

There are only 70 unique objects serialized, so it looks like From is creating 2 objects per (which seems high) and To is creating 4 objects per.  Since To is map based I could maybe understand there being a higher constant factor per type (both slice and map), but that's still a lot of overhead I wouldn't expect.

Is "to" creating a string for each map key?  We shouldn't have to allocate a new map key for each field since the string should be immutable.

wojtek-t committed:
Better tests for unrecognized types

wojtek-t commented:
Regarding the further improvements:

> It might be map[string]string converted from map[string]interface{}, and that instead of doing a generic iteration with reflect, you should be special casing map[string]string (when it's the destination type) with a special for loop that sets a map[string]string). That might be a big win.

It's pretty much this commented code: 
https://github.com/kubernetes/kubernetes/pull/39418/files#diff-505d11c5374bda934c612fa3befcae46R372

I tried this recently, but something didn't work for me in the api tests. But when using it in benchmarking, it didn't give any significant improvement (it was < 5%)

> Is "to" creating a string for each map key? We shouldn't have to allocate a new map key for each field since the string should be immutable.

I hope not. The fieldInfo.nameValue is supposed to handle it:
https://github.com/kubernetes/kubernetes/pull/39418/files#diff-505d11c5374bda934c612fa3befcae46R488

I will try to dig deeper once this PR is merged.

wojtek-t commented:
@smarterclayton - PTAL

wojtek-t commented:
FYI - as an experiment I did the following changes to "ToUnstructured":
- I added "return nil" for handling all slices and maps (I'm not even recursing into them)
- I changed handling a nested struct, to add fields to the current map[string]interface{} intead of creating a new one
For that and your example, I'm getting the following result:
```
	unstructured_test.go:248: map[string]interface {}{"restartPolicy":"Always", "hostIPC":false, "hostname":"", "qosClass":"", "kind":"Pod", "apiVersion":"v1", "serviceAccount":"", "hostPID":false, "namespace":"default", "hostNetwork":false, "schedulername":"", "nodeName":"e2e-test-wojtekt-master", "generateName":"", "selfLink":"/api/v1/namespaces/default/pods/etcd-server-e2e-test-wojtekt-master", "subdomain":"", "hostIP":"", "uid":"a671734a-e8e5-11e4-8fde-42010af09327", "clusterName":"", "podIP":"", "serviceAccountName":"", "message":"", "generation":0, "dnsPolicy":"ClusterFirst", "phase":"Running", "reason":"", "name":"etcd-server-e2e-test-wojtekt-master", "resourceVersion":"22"}
```
And even for that have 61 allocations:
```
BenchmarkToFromUnstructured-12           	   10000	    232281 ns/op	    5487 B/op	      61 allocs/op
```

There are 26 fields here, so it seems we have ~2 allocations per field in a map. So it seems I'm doing something wrong, but looking into code I don't really know why..

wojtek-t commented:
I'm really confused. the following benchmark works as I would expect it to work:
```
func BenchmarkXxx(b *testing.B) {
	b.ResetTimer()

	q := []string{"abc", "qwe", "wer", "ert", "tyu", "sdf", "hgf", "lkj", "mnb", "cxz"}
	t := reflect.TypeOf(map[string]string{})
	c := make([]reflect.Value, 10, 10)
	for i := 0; i < len(q); i++ {
		c[i] = reflect.ValueOf(q[i])
	}

	for i := 0; i < b.N; i++ {
		x := reflect.MakeMap(t)
		for i := 0; i < len(q); i++ {
			x.SetMapIndex(c[i], c[i])
		}
		y := x.Interface().(map[string]string)
		for i := 0; i < len(q); i++ {
			if y[q[i]] != q[i] {
				b.Fatalf("dupa")
			}
		}
	}
	b.StopTimer()
}

BenchmarkXxx-12           	  200000	      8593 ns/op	     918 B/op	       3 allocs/op
```

So I have completely no idea, why we have some many allocations coming from:
```
subv = fv
...
dv.SetMapIndex(fieldInfo.nameValue, subv)
```
lines in structToUnstructured function...

wojtek-t committed:
Simple optimization in structToUnstructured

wojtek-t commented:
@smarterclayton - I figured out one simple optimization we can do - see the last commit.

smarterclayton commented on staging/src/k8s.io/apimachinery/pkg/conversion/unstructured/converter_test.go:
```diff
@@ -179,3 +202,71 @@ func TestRoundTrip(t *testing.T) {
+		obj  runtime.Object
+	}{
+		{
+			data: "{\"da\":[3.0,\"3.0\",null]}",
```
I'm not doing a good job of explaining what the issue is :)

1. Imagine a destination struct v1.Pod
2. A test that verifies that `{"metadata":{"labels": {"a": null}}}`, when unmarshalled into v1.Pod via JSON or via JSON -> map[string]interface{} -> UnstructuredTo, results in the exact same object or error
3. Test other variations, like `{"spec": {"containers": [null]}}`, `{"spec": {"containers": [1]}}`, etc.  UnstructuredTo *must* be a drop in replacement for JSON.Unmarshal.

wojtek-t replied:
OK - I hope that this time I understood correctly.
I changed the test to check if:
1) serialized json -> object
2) serialized json -> map[string]interface{} -> object
produce the same output

I also added a bunch of other test cases.

I hope that this time is what you wanted.

smarterclayton commented on staging/src/k8s.io/apimachinery/pkg/conversion/unstructured/converter_test.go:
```diff
@@ -52,11 +53,38 @@ type C struct {
+}
+
+type F struct {
+	A string `json:"fa"`
```
Add int, float, []string, []int, and []bool, and then be sure to try to send mismatching values to each (int, float, bool -> string, []int -> []string, int -> []string, int -> []int, etc).

wojtek-t replied:
I added ~20 more cases, and it seems 3 of them are not passing. I will try to debug them when I have some cycles.

wojtek-t replied:
@smarterclayton - I just checked it, and it seems some of those doesn't have any chance to work.
Basically, if the input is incorrect, in some cases just doing:
1. string -> json.Unmarshal - > object
vs
2. string -> json.Unmarshal -> map[string]interface{} ->json.Marshal -> []byte -> json.Unmarshal -> object

Returns different results (i.e. the first one is error, the latter is not).

That said, since the "FromUnstructured" works on data from 2nd scenario, it doesn't have a way to error (so we cannot expect it to return the error).

Though, there are still some differences, which I will debug further next week.

wojtek-t replied:
@smarterclayton - WDYT?

smarterclayton commented:
To be really sure we're safe, I want to do something like KUBE_CACHE_MUTATION_DETECTOR.  By default, we call the new code path in the patch function.  However, in the gce test suites we set KUBE_PATCH_CONVERSION_DETECTOR=true and in your code path we call both and compare the output.  If they mismatch, panic.  That would give me the confidence that even if there's some edge case we're missing here, that the GCE tests flush it out.

wojtek-t commented:
> To be really sure we're safe, I want to do something like KUBE_CACHE_MUTATION_DETECTOR. By default, we call the new code path in the patch function. However, in the gce test suites we set KUBE_PATCH_CONVERSION_DETECTOR=true and in your code path we call both and compare the output. If they mismatch, panic. That would give me the confidence that even if there's some edge case we're missing here, that the GCE tests flush it out.

@smarterclayton - I agree it makes sense. Though, can we make it in a separate PR (before we switch it on in k8s code). This PR is not starting to use it, so it shouldn't break anything.

smarterclayton commented:
Yes, it's fine to merge this and before we introduce it in the code path
have the verification path.

On Fri, Feb 3, 2017 at 10:09 AM, Wojciech Tyczynski <
notifications@github.com> wrote:

> To be really sure we're safe, I want to do something like
> KUBE_CACHE_MUTATION_DETECTOR. By default, we call the new code path in the
> patch function. However, in the gce test suites we set
> KUBE_PATCH_CONVERSION_DETECTOR=true and in your code path we call both
> and compare the output. If they mismatch, panic. That would give me the
> confidence that even if there's some edge case we're missing here, that the
> GCE tests flush it out.
>
> @smarterclayton <https://github.com/smarterclayton> - I agree it makes
> sense. Though, can we make it in a separate PR (before we switch it on in
> k8s code). This PR is not starting to use it, so it shouldn't break
> anything.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/39418#issuecomment-277270902>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_pwm-ptvOZtFlvBe92sEvpcxRLfKpks5rY0MWgaJpZM4Laj1V>
> .
>

wojtek-t commented:
@smarterclayton - I fixed the code so that all test are passing now (I also added few more).

PTAL - this should hopefully be ready to do.

smarterclayton commented:
@k8s-bot kops aws e2e test this

smarterclayton commented:
/lgtm

wojtek-t commented:
Trivial rebase (vendor/BUILD). Reapplying labels.
